tuxedo-parachain-core/src/validate_block.rs

```diff
@@ -0,0 +1,435 @@
+        // Seems like we could call the existing collect_collation_info api to get this information here
+        // instead of FRAME's approach of tightly coupling to pallet parachain system.
+        // That would mean less duplicated code as well as a more flexible validate block macro.
+
+        // Get the relay parent number out of storage so we can advance the hrmp watermark
+        let hrmp_watermark = RelayParentNumberStorage::get();
+
+        ValidationResult {
+            head_data,
+            new_validation_code: None,
+            upward_messages: Default::default(),
+            processed_downward_messages: 0,
+            horizontal_messages: Default::default(),
+            hrmp_watermark,
+        }
+    })
+}
+
+/// Extract the [`ParachainInherentData`] from a parachain block.
+/// The data has to be extracted from the extrinsics themselves.
+/// I want the runtime to expose a method to do this, and I also want it to
+/// be nice and flexible by searching for the right transactions.
+/// For now I have a hacky implementation that assumes the parachain inherent is last
+fn extract_parachain_inherent_data<B, V, C>(block: &B) -> ParachainInherentData
+where
+    B: BlockT<Extrinsic = Transaction<V, C>>,
+    // Consider an alternative way to express the bounds here:
+    // Transaction<V, C>: Extrinsic
+    V: TypeInfo + Verifier + 'static,
+    C: TypeInfo + ConstraintChecker<V> + 'static,
+{
+    // The commented stuff is Basti's algo.
+    // It is nicer than my hack because it searches the transactions,
+    // But it is still not good enough because it lived right here in this file as
+    // opposed to with the runtime.
+    // FIXME https://github.com/Off-Narrative-Labs/Tuxedo/issues/146
+
+    // block
+    // 	.extrinsics()
+    // 	.iter()
+    // 	// Inherents are at the front of the block and are unsigned.
+    // 	//
+    // 	// If `is_signed` is returning `None`, we keep it safe and assume that it is "signed".
+    // 	// We are searching for unsigned transactions anyway.
+    // 	.take_while(|e| !e.is_signed().unwrap_or(true))
+    // 	.filter_map(|e| e.call().is_sub_type())
+    // 	.find_map(|c| match c {
+    // 		crate::Call::set_validation_data { data: validation_data } => Some(validation_data),
+    // 		_ => None,
+    // 	})
+    // 	.expect("Could not find `set_validation_data` inherent")
+
+    block
+        .extrinsics()
+        .iter()
+        .take_while(|&e| !e.is_signed().unwrap_or(true))
+        .collect::<Vec<_>>()
+        .last()
+        .expect("There should be at least one inherent extrinsic which is the parachain inherent.")
+        .outputs
+        .get(0)
+        .expect("Parachain inherent should be first and should have exactly one output.")
+        .payload
+        .extract::<ParachainInherentDataUtxo>()
+        .expect("Should decode to proper type based on the position in the block.")
+        .into()
```

Comment:
This is shorter, sexier, and probably works - please test it!
```suggestion
    block
        .extrinsics()
        .iter()
        .take_while(|&e| !e.is_signed().unwrap_or(true))
        .find_map(|&e| {
            e.outputs
                .get(0)
                .map(|o| o.payload.extract::<ParachainInherentDataUtxo>())
                .flatten()
        })
        .expect("There should be at least one inherent extrinsic which is the parachain inherent.")
```

Reply:
I'd prefer to defer this change at least until we tackle #146